Fix tablet settings incorrectly reporting invalid area

peppy opened the pull request:
Was ready to spend some time figuring out the particular reasoning for floating point issues, but this fix seems the simplest (and I haven't been able to break it since).

Closes https://github.com/ppy/osu/issues/14570

peppy committed:
Expose more pieces of `TabletSettings`

peppy committed:
Add test coverage of failing validity checks

peppy committed:
Fix bounds check running too early causing tablet area to show incorrect validity

bdach commented:
Still fails here with 270 degree rotation (which is not tested):

![2021-09-04-131727_517x854_scrot](https://user-images.githubusercontent.com/20418176/132092596-f01f0cd5-59be-4c97-ae81-1099e7edeab1.png)

Also the test is not testing properly because it is not waiting for the transforms to end, which means that if the transforms are interrupted midway by rotation changing then the `IsWithinBounds` flag will never actually change value.

bdach commented:
I'll leave the updated test code here for convenience:

```csharp
        [Test]
        public void TestValidAfterRotation()
        {
            AddAssert("area valid", () => settings.AreaSelection.IsWithinBounds);

            AddStep("rotate 90", () => tabletHandler.Rotation.Value = 90);
            AddUntilStep("wait for transforms", () => settings.AreaSelection.ChildrenOfType<Container>().All(c => !c.Transforms.Any()));
            AddAssert("area valid", () => settings.AreaSelection.IsWithinBounds);

            AddStep("rotate 180", () => tabletHandler.Rotation.Value = 180);
            AddUntilStep("wait for transforms", () => settings.AreaSelection.ChildrenOfType<Container>().All(c => !c.Transforms.Any()));
            AddAssert("area valid", () => settings.AreaSelection.IsWithinBounds);

            AddStep("rotate 270", () => tabletHandler.Rotation.Value = 270);
            AddUntilStep("wait for transforms", () => settings.AreaSelection.ChildrenOfType<Container>().All(c => !c.Transforms.Any()));
            AddAssert("area valid", () => settings.AreaSelection.IsWithinBounds);

            AddStep("rotate 360", () => tabletHandler.Rotation.Value = 360);
            AddUntilStep("wait for transforms", () => settings.AreaSelection.ChildrenOfType<Container>().All(c => !c.Transforms.Any()));
            AddAssert("area valid", () => settings.AreaSelection.IsWithinBounds);

            AddStep("rotate 0", () => tabletHandler.Rotation.Value = 0);
            AddUntilStep("wait for transforms", () => settings.AreaSelection.ChildrenOfType<Container>().All(c => !c.Transforms.Any()));
            AddAssert("area valid", () => settings.AreaSelection.IsWithinBounds);
        }
```

bdach reviewed (changes requested):
see above

peppy committed:
Add extended tests

peppy commented:
I did actually test with 270 degrees and it didn't repro locally (even headless)... until I made my window much smaller.

Will take another look, the 1px tolerance is obviously not correct after downscaling.

bdach commented:
Yep, the above test also passed headless for me, but not in the actual visual test.

peppy committed:
Update check to inflate in the correct direct

Also handles previously unhandled edge cases by comparing all four
corners, instead of only two.

peppy committed:
Add test coverage of invalid cases too

peppy commented:
I've *hopefully* resolved this now. Turns out the previous transform attachment was actually fine where it was, and the issue was incorrect inflation logic when rotation was non-zero.

bdach commented on osu.Game/Overlays/Settings/Sections/Input/TabletAreaSelection.cs:
```diff
@@ -169,12 +174,24 @@ private void checkBounds()
+            var usableAreaQuad = Quad.FromRectangle(usableAreaContainer.ScreenSpaceDrawQuad.AABBFloat);
+
+            var matrix = Matrix3.Identity;
+            MatrixExtensions.TranslateFromLeft(ref matrix, usableAreaQuad.Centre);
+            MatrixExtensions.RotateFromLeft(ref matrix, MathUtils.DegreesToRadians(rotation.Value));
+            MatrixExtensions.TranslateFromLeft(ref matrix, -usableAreaQuad.Centre);
```
Out of curiosity, was there a particular reason why you used the `FromLeft` functions here? I was a little confused because the order of operations seemed backward (as I - seemingly correctly - interpreted that what this was doing is moving the quad so that its centre is at (0,0), then rotating it and moving the centre back where it was), but I got there in the end and understood that these calls were doing the transformation composition by multiplying *from the left*. But using `FromRight` with the order swapped is just as valid and is maybe a smidge more readable.

Either way this looks to work now, so I'm not really fussed either way. :+1:

peppy replied:
I was just basing this off what we already did with `DrawNode` manipulation. No real reason.

PercyDan54 commented on osu.Game.Tests/Visual/Settings/TestSceneTabletSettings.cs:
```diff
@@ -17,29 +18,93 @@ namespace osu.Game.Tests.Visual.Settings
+            AddStep("rotate 0", () => tabletHandler.Rotation.Value = 0);
+            ensureValid();
+
+            AddStep("rotate 0", () => tabletHandler.Rotation.Value = 45);
```
Is this a leftover when copy-pasting?
```suggestion
            AddStep("rotate 45", () => tabletHandler.Rotation.Value = 45);
```

bdach replied:
i had a change fixing this but forgot to push it. will do the next time i can, unless preempted.

peppy committed:
Fix incorrect test step name

Co-authored-by: PercyDan <50285552+PercyDan54@users.noreply.github.com>

bdach commented:
Found yet another broken case, unfortunately:

![2021-09-05-144114_464x345_scrot](https://user-images.githubusercontent.com/20418176/132127174-29b18120-4f7a-40e5-9cf1-503022f0d129.png)

AABB cannot be used because both of the containers are rotated individually and the AABB does not work on a "global" level, just for a given quad. (So for example the fact that `usableAreaContainer` is always upright cannot be relied upon.)

I tried a few different things but almost every one of them was torpedoed by `Quad.Contains()` being _wayyy_ too sensitive to floating point fluctuations. Even scaling the usable area container down by 5% didn't help.

bdach reviewed (changes requested):
as above :(

peppy committed:
Avoid using SSDQ for validity computation

peppy committed:
Add test coverage of sharper aspect ratio

peppy commented:
Rewritten to not use SSDQ. Seems better. Please don't find any more edge cases.

bdach commented:
I've tried for a good fifteen minutes and found nothing. Although I understand what the code is doing I don't get why it's so much less sensitive to fp inaccuracies than whatever I was trying.

Either way, this seems to be behaving well enough to ship. :+1:

bdach committed:
Merge branch 'master' into tablet-area-validity-fix
